### ISSUE TITLE
Deserialization of strings is incorrect

### ISSUE THREAD

**Issue by sumanth-nirmal** [eclipse-cyclonedds/cyclonedds-cxx#117]
## Required information

**Operating system:**
Ubuntu 18.04 LTS

**Cyclone version:**
- cyclonedds - master, [779c58](https://github.com/eclipse-cyclonedds/cyclonedds/commit/779c58206d6240a2cc405eece9a66bf70b97cd00)
- cyclonedds-cxx - master, 624da660d9d2c725555bd65bf5ff5c879dbee77b

**Observed result or behavior:**

The deserialization doesn't happen correctly with the sequence/array of strings

**Detailed reproducer**

When using the following IDL file

```hpp
module my_msg {
  module dds {
    typedef string string__3[3];
    struct msg {
      string__3 string_values;
    };
  };
};
```
in the code as below:
```cpp
TEST(tooch, deserialize_string)
{
  using DDSType = my_msg::dds::msg;

  DDSType deserialized_msg{};

  // serialized buffer
  uint8_t serialized_buf[] =
  {0x01, 0x00, 0x01, 0x00, 0x0f, 0x00, 0x00, 0x00, 0x43, 0x6f, 0x6d, 0x70, 0x6c, 0x65, 0x78, 0x20,
    0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x31, 0x00, 0x00, 0x0f, 0x00, 0x00, 0x00, 0x43, 0x6f, 0x6d,
    0x70, 0x6c, 0x65, 0x78, 0x20, 0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x32, 0x00, 0x00, 0x0f, 0x00,
    0x00, 0x00, 0x43, 0x6f, 0x6d, 0x70, 0x6c, 0x65, 0x78, 0x20, 0x48, 0x65, 0x6c, 0x6c, 0x6f,
    0x33, 0x00, 0x00};

  //deserialize the buffer
  org::eclipse::cyclonedds::core::cdr::basic_cdr_stream cdr_streamer;
  cdr_streamer.set_buffer(serialized_buf + 4); // ignore the header
  org::eclipse::cyclonedds::core::cdr::read(cdr_streamer, deserialized_msg);
  EXPECT_FALSE(cdr_streamer.abort_status());
  EXPECT_STREQ(deserialized_msg.string_values()[0].c_str(), "Complex Hello1");
  EXPECT_STREQ(deserialized_msg.string_values()[1].c_str(), "Complex Hello2");
  EXPECT_STREQ(deserialized_msg.string_values()[2].c_str(), "Complex Hello3");
}
```
The test doesn't work (I have serialized the data by hand)
I guess this is because of 2 reasons
1. The `string_length - 1` here https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/master/src/ddscxx/include/org/eclipse/cyclonedds/core/cdr/basic_cdr_ser.hpp#L262 could lead to a huge number when the string is empty, which causes problems
2. This alignment here https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/master/src/ddscxx/include/org/eclipse/cyclonedds/core/cdr/basic_cdr_ser.hpp#L267 doesn't seem correct, it is not always aligned by 1, correct?

**Expected result or behaviour:**

The deserialization should work for all types in all scenarios

**Comment by reicheratwork** [comment]
1. getting a length field of 0 should never appear in valid CDR data, as the length field for a string includes the terminating NULL (CORBA v3.3 9.3.2.7) so this is always a minimum of 1, but since the reading has to be able to correctly handle even incorrect CDR streams, a check on `string_length > 0` is a necessary addition, the fix for this is already in PR #112 
2. this is indeed an error, as the CDR stream that you provide is correct, there is an issue with the entity being read with the cursor from before the alignment step is done, i will work on a fix for this

The stream you provided is in little-endian format, which is not reflected in the endianness flag set in the message header (byte offset 1 = `0x0`)

**Comment by sumanth-nirmal** [comment]
> The stream you provided is in little-endian format, which is not reflected in the endianness flag set in the message header (byte offset 1 = 0x0)

That's because of my incorrect hand serialization, but the header should actually reflect little endianness (offset 1 = 0x1)
